File: apps/frontend/src-tauri/src/models/onboarding.rs
//! Onboarding types for the AgentPod Tauri app
//! These models mirror the Management API onboarding types

use serde::{Deserialize, Serialize};

/// Onboarding session status
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum OnboardingStatus {
    Pending,
    Started,
    Gathering,
    Generating,
    Applying,
    Completed,
    Skipped,
    Failed,
}

/// Onboarding requirements gathered from user
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OnboardingRequirements {
    pub project_type: String,
    pub project_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub primary_language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frameworks: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub build_tools: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub testing_framework: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub linter: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub formatter: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preferred_model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preferred_small_model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub coding_style: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_instructions: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub additional_context: Option<serde_json::Value>,
}

/// Agent definition in generated config
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentDefinition {
    pub name: String,
    pub description: String,
    pub content: String,
}

/// Command definition in generated config
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CommandDefinition {
    pub name: String,
    pub description: String,
    pub content: String,
}

/// MCP server definition in generated config
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpServerDefinition {
    pub name: String,
    pub command: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub args: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env: Option<std::collections::HashMap<String, String>>,
}

/// Generated configuration from onboarding
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GeneratedConfig {
    pub settings: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agents_md: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agents: Option<Vec<AgentDefinition>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub commands: Option<Vec<CommandDefinition>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mcp_servers: Option<Vec<McpServerDefinition>>,
}

/// Onboarding session
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OnboardingSession {
    pub id: String,
    pub user_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sandbox_id: Option<String>,
    pub status: OnboardingStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gathered_requirements: Option<OnboardingRequirements>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generated_config: Option<GeneratedConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected_model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected_small_model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<String>,
}

/// Response wrapper for onboarding session
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OnboardingSessionResponse {
    pub session: OnboardingSession,
}

/// Result of applying onboarding config
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyConfigResult {
    pub success: bool,
    pub files_written: Vec<String>,
    pub session: OnboardingSession,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Input for completing onboarding
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompleteOnboardingInput {
    pub config: GeneratedConfig,
}

/// Input for resetting onboarding
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResetOnboardingInput {
    #[serde(default)]
    pub preserve_models: bool,
}

/// Input for creating onboarding session
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateOnboardingInput {
    pub sandbox_id: String,
}

File: apps/console/src/lib/api/onboarding.ts
/**
 * Onboarding API
 * 
 * Tauri invoke wrappers for the Management API's onboarding endpoints.
 * This module provides typed wrappers around Tauri commands.
 */

import { invoke } from "./tauri";

// =============================================================================
// Types
// =============================================================================

export type OnboardingStatus =
  | "pending"
  | "started"
  | "gathering"
  | "generating"
  | "applying"
  | "completed"
  | "skipped"
  | "failed";

export interface OnboardingRequirements {
  projectType: string;
  projectName: string;
  projectDescription?: string;
  primaryLanguage?: string;
  frameworks?: string[];
  buildTools?: string[];
  testingFramework?: string;
  linter?: string;
  formatter?: string;
  preferredModel?: string;
  preferredSmallModel?: string;
  codingStyle?: string;
  customInstructions?: string;
  additionalContext?: Record<string, unknown>;
}

export interface GeneratedConfig {
  settings: Record<string, unknown>;
  agentsMd?: string;
  agents?: Array<{
    name: string;
    description: string;
    content: string;
  }>;
  commands?: Array<{
    name: string;
    description: string;
    content: string;
  }>;
  mcpServers?: Array<{
    name: string;
    command: string;
    args?: string[];
    env?: Record<string, string>;
  }>;
}

export interface OnboardingSession {
  id: string;
  userId: string;
  sandboxId: string | null;
  status: OnboardingStatus;
  projectType: string | null;
  projectName: string | null;
  projectDescription: string | null;
  gatheredRequirements: OnboardingRequirements | null;
  generatedConfig: GeneratedConfig | null;
  selectedModel: string | null;
  selectedSmallModel: string | null;
  errorMessage: string | null;
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
}

export interface ApplyConfigResult {
  success: boolean;
  filesWritten: string[];
  session: OnboardingSession;
  error?: string;
}

// =============================================================================
// Onboarding API Functions (via Tauri invoke)
// =============================================================================

/**
 * Get onboarding session for a sandbox
 */
export async function getOnboardingSession(sandboxId: string): Promise<OnboardingSession | null> {
  return invoke<OnboardingSession | null>("get_onboarding_session", { sandboxId });
}

/**
 * Get onboarding session by ID
 */
export async function getOnboardingSessionById(sessionId: string): Promise<OnboardingSession> {
  return invoke<OnboardingSession>("get_onboarding_session_by_id", { sessionId });
}

/**
 * Create a new onboarding session for a sandbox
 */
export async function createOnboardingSession(sandboxId: string): Promise<OnboardingSession> {
  return invoke<OnboardingSession>("create_onboarding_session", { sandboxId });
}

/**
 * Start an onboarding session (changes status from pending to started)
 */
export async function startOnboarding(sessionId: string): Promise<OnboardingSession> {
  return invoke<OnboardingSession>("start_onboarding", { sessionId });
}

/**
 * Skip onboarding for a session
 */
export async function skipOnboarding(sessionId: string): Promise<OnboardingSession> {
  return invoke<OnboardingSession>("skip_onboarding", { sessionId });
}

/**
 * Complete onboarding with generated config
 */
export async function completeOnboarding(
  sessionId: string,
  config: GeneratedConfig
): Promise<OnboardingSession> {
  return invoke<OnboardingSession>("complete_onboarding", { sessionId, config });
}

/**
 * Apply onboarding configuration to the sandbox
 */
export async function applyOnboardingConfig(
  sessionId: string,
  reload = true
): Promise<ApplyConfigResult> {
  return invoke<ApplyConfigResult>("apply_onboarding_config", { sessionId, reload });
}

/**
 * Reset onboarding session to start over
 */
export async function resetOnboarding(
  sessionId: string,
  preserveModels = false
): Promise<OnboardingSession> {
  return invoke<OnboardingSession>("reset_onboarding", { sessionId, preserveModels });
}
